rsmtool/utils.py

```diff
@@ -506,62 +506,228 @@ def agreement(score1, score2, tolerance=0):
-    scale. Additional implementations may be added in the future.
+    scale.
     """
+    y_true_avg = np.mean(y_true)
+    y_pred_avg = np.mean(y_pred)
+    numerator = y_true_avg - y_pred_avg
+
     method = method.lower()
-    if method == 'williamson':
-        numerator = mean_system_score - mean_human_score
-        denominator = np.sqrt((population_system_score_sd**2 +
-                               population_human_score_sd**2) / 2)
+    if method == 'unpooled':
+        denominator = np.std(y_true, ddof=ddof)
+    elif method == 'pooled':
+        denominator = np.sqrt((np.std(y_true, ddof=ddof)**2 +
+                               np.std(y_pred, ddof=ddof)**2) / 2)
+    elif method == 'johnson':
+        if population_y_true_sd is None:
+            raise ValueError("If `method='johnson'`, then `population_y_true_sd` "
+                             "must be provided.")
+        denominator = population_y_true_sd
+    elif method == 'williamson':
+        if population_y_true_sd is None or population_y_pred_sd is None:
+            raise ValueError("If `method='williamson'`, both `population_y_true_sd` "
+                             "and `population_y_pred_sd` must be provided.")
+        numerator = y_pred_avg - y_true_avg
+        denominator = np.sqrt((population_y_true_sd**2 +
+                               population_y_pred_sd**2) / 2)
     else:
-        raise ValueError("Currently, only the 'williamson' SMD method is "
-                         "supported. You specified: {}".format(method))
+        possible_methods = {"'unpooled'", "'pooled'", "'johnson'", "'williamson'"}
+        raise ValueError("The available methods are {{{}}}; you selected {}."
+                         "".format(', '.join(possible_methods), method))
 
     # if the denominator is zero, then return NaN as the SMD
     smd = np.nan if denominator == 0 else numerator / denominator
     return smd
 
 
+def difference_of_standardized_means(y_true,
+                                     y_pred,
+                                     population_y_true_mn=None,
+                                     population_y_pred_mn=None,
+                                     population_y_true_sd=None,
+                                     population_y_pred_sd=None,
+                                     ddof=1):
+    """
+    Calculate the difference of standardized means.
+
+    Parameters
+    ----------
+    y_true : array-like
+        The true score for the group or subgroup.
+    y_pred : array-like
+        The predicted score for the group or subgroup.
+        The predicted scores.
+    population_y_true_mn : float or None, optional
+        The population true score mean.
+        When the DSM is being calculated for a subgroup,
+        this should be the mean for the whole
+        population.
+        Defaults to None.
+    population_y_pred_mn : float or None, optional
+        The predicted score mean.
```

Comment:
Nitpick: I am not sure about the use case where the z-scores are computed *without* the population mean as the DSM value in this case will always  be 0, right?

Reply:
Yeah, it will always be zero. The only reason I allow this is, again, to follow `sklearn` conventions, where the only required arguments for a metric function should be `y_true` and `y_pred`, and all other arguments should be optional.

I'm all right changing it, though. Just let me know what you think makes the most sense.

Reply:
I agree. May be add a a warning if function is used as API by somebody who hasn't thought this through?